### PR TITLE
Added C program to find inorder predecessor and successor in a BST

### PR DESCRIPTION
**Pull request by Tanya170620**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### Description

It is a menu driven program which helps to find the inorder predecessor and successor in a BST.
The user first create their desired BST and then can find the inorder predecessor and successor of a node.

### Checklist

- [Yes] I've read the contribution guidelines.
- [Yes] I've checked the issue list before deciding what to submit.
- [Yes] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

To find inorder successor and predecessor in BST #3388

### CODE REVIEW

File: C/ds/inorder_predecessor_successor.c [hunk 1]
```diff
@@ -0,0 +1,241 @@
+/*Inorder successor of a node is the next node in inorder traversal of the BST.
+There is no inorder successor of the last node in inorder traversal of the BST.
+Inorder predecessor of a node is the previous node in inorder traversal of the BST.
+There is no inorder predecessor of the first node in inorder traversal of the BST.
+*/
+#include<stdio.h>
```
Comment (ankitaggarwal23) [inline C/ds/inorder_predecessor_successor.c]: Add a new line after description

File: C/ds/inorder_predecessor_successor.c [hunk 2]
```diff
@@ -0,0 +1,241 @@
+/*Inorder successor of a node is the next node in inorder traversal of the BST.
+There is no inorder successor of the last node in inorder traversal of the BST.
+Inorder predecessor of a node is the previous node in inorder traversal of the BST.
+There is no inorder predecessor of the first node in inorder traversal of the BST.
+*/
+#include<stdio.h>
+#include<stdlib.h>
+//structure of tree
```
Comment (ankitaggarwal23) [inline C/ds/inorder_predecessor_successor.c]: Remove the redundant comments

File: C/ds/inorder_predecessor_successor.c [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+#include<stdio.h>
+#include<stdlib.h>
+//structure of tree
+struct tree
+{
+    struct tree *lptr;
+    int data;
+    struct tree *rptr;
+};
+//Insert function of BST
```
Comment (ankitaggarwal23) [inline C/ds/inorder_predecessor_successor.c:17]: Add a new line above the comments

File: C/ds/inorder_predecessor_successor.c [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    }
+    else
+    {
+        if((*b)->data>n)
+        insert(&((*b)->lptr),n);
+        else
+        insert(&((*b)->rptr),n);
+    }
+}
+//Inorder display function using recursion
```
Comment (ankitaggarwal23) [inline C/ds/inorder_predecessor_successor.c:36]: Add a new line above the comments

File: C/ds/inorder_predecessor_successor.c [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+void inorder(struct tree *b)
+{
+    if(b!=NULL)
+    {
+        inorder(b->lptr);
+        printf("%d  ",b->data);
+        inorder(b->rptr);
+    }
+}
+//Function to find inorder predecessor of a node
```
Comment (ankitaggarwal23) [inline C/ds/inorder_predecessor_successor.c:47]: Add a new line above the comments

File: C/ds/inorder_predecessor_successor.c [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        {
+            xpre=xpre->rptr;
+        }
+        printf("Predecessor of %d is %d\n",node,xpre->data);
+        return 0;
+    }
+    else if(flag==1)
+    printf("Predecessor of %d is %d\n",node,pre);
+    else
+    printf("Predecessor of %d does not exist\n",node);
```
Comment (ankitaggarwal23) [inline C/ds/inorder_predecessor_successor.c]: Give proper one tab space inside the body of the condition.

File: C/ds/inorder_predecessor_successor.c [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+}
+//Function to find inorder successor of a node
```
Comment (ankitaggarwal23) [inline C/ds/inorder_predecessor_successor.c:87]: Add a new line above the comments

File: C/ds/inorder_predecessor_successor.c [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    while(x!=NULL && x->data!=node)
+    {
+        if(x->data>node)
+        {
+            f=1;
+            xsucc=x->data;
+            x=x->lptr;
+        }
+        else
+        x=x->rptr;
```
Comment (ankitaggarwal23) [inline C/ds/inorder_predecessor_successor.c]: Give proper one tab space inside the body of the condition.

File: C/ds/inorder_predecessor_successor.c [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+        printf("Node not present in tree\n");
+        return 0;
+    }
+    if(x->rptr!=NULL)
+    {
+        x=x->rptr;
+        while(x->lptr!=NULL)
+        x=x->lptr;
+        printf("Successor of %d is %d\n",node,x->data);
+        return 0;
```
Comment (ankitaggarwal23) [inline C/ds/inorder_predecessor_successor.c]: Give proper one tab space inside the body of the condition.

File: C/ds/inorder_predecessor_successor.c [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+    }
+    else if(f==1)
+    printf("Successor of %d is %d\n",node,xsucc);
+    else
+    printf("Successor of %d does not exist\n",node);
+}
```
Comment (ankitaggarwal23) [inline C/ds/inorder_predecessor_successor.c]: Give proper one tab space inside the body of the condition.

File: C/ds/inorder_predecessor_successor.c [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+            {
+                printf("Enter data: ");
+                scanf("%d",&n);
+                insert(&root,n);
+                break;
+            }
+            case 2:
+            {
+                if(root==NULL)
+                printf("Tree is empty\n");
```
Comment (ankitaggarwal23) [inline C/ds/inorder_predecessor_successor.c]: Give proper one tab space inside the body of the condition.

File: C/ds/inorder_predecessor_successor.c [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+                    printf("In-order Display: ");
+                    inorder(root);
+                    printf("\n");
+                }
+                break;
+            }
+            case 3:
+            {
+                if(root==NULL)
+                printf("Tree is empty\n");
```
Comment (ankitaggarwal23) [inline C/ds/inorder_predecessor_successor.c]: Give proper one tab space inside the body of the condition.

File: C/ds/inorder_predecessor_successor.c [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+                    printf("Enter the node for which you want inorder predecessor: ");
+                    scanf("%d",&n);
+                    in_pre(root,n);
+                }
+                break;
+            }
+            case 4:
+            {
+                if(root==NULL)
+                printf("Tree is empty\n");
```
Comment (ankitaggarwal23) [inline C/ds/inorder_predecessor_successor.c]: Give proper one tab space inside the body of the condition.

File: C/ds/inorder_predecessor_successor.c [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+Do you want to continue:(Y/y) y
+
+Menu:
+1.Insert                                2.Inorder display
+3.Inorder Predecessor of a node         4.Inorder Successor of a node
+Enter your choice: 4
+Enter the node for which you want inorder successor: 5
+Successor of 5 is 9
+Do you want to continue:(Y/y) n
+*/
```
Comment (ankitaggarwal23) [inline C/ds/inorder_predecessor_successor.c]: - Remove the extra test case 3  5  9  14 (Use only four values as sample input)
- Add Time and Space Complexity at the end of the Code

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]

File: C/ds/inorder_predecessor_successor.c [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+    while(x!=NULL && x->data!=node)
+    {
+        if(x->data>node)
+        {
+            f=1;
+            xsucc=x->data;
+            x=x->lptr;
+        }
+        else
+        x=x->rptr;
```
Comment (ankitaggarwal23) [inline C/ds/inorder_predecessor_successor.c]: Give proper one tab space inside the body of the condition.

File: C/ds/inorder_predecessor_successor.c [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+    if(x==NULL)
+    {
+        printf("Node not present in tree\n");
+        return 0;
+    }
+    if(x->rptr!=NULL)
+    {
+        x=x->rptr;
+        while(x->lptr!=NULL)
+        x=x->lptr;
```
Comment (ankitaggarwal23) [inline C/ds/inorder_predecessor_successor.c]: Give proper one tab space inside the body of the condition.

File: C/ds/inorder_predecessor_successor.c [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+    return 0;
+}
+/*
+Time Complexity:
+Insertion - O(n)
+Inorder display - O(n)
+Inorder predecessor function - O(h)
+Inorder successor function - O(h)
+
+Space Complexity: O(n)
```
Comment (ankitaggarwal23) [inline C/ds/inorder_predecessor_successor.c]: Shift this after I/P and O/P. What is O(h) I guess it is O(n) kindly change.

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]

**Comment by Tanya170620** [comment]
@ankitaggarwal23  I'm sorry as I did not know that if I have an existing PR which was not closed yet and if I make another PR for another issue then it will be reflected in all my PRs. 
Should I close all the PRs and make a fresh PR for this one or what do you suggest?

**Comment by Tanya170620** [comment]
Should I make a new PR for this issue or is it fine right now?
@ankitaggarwal23 @gat786 @bhav09

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Use a formatter to format the code. Rest looks good!

**Review by HarshCasper — APPROVED** [review]

**Pull request merged**